src/business_logic/transaction/fee.rs

```diff
@@ -0,0 +1,124 @@
+use super::error::TransactionError;
+use crate::{
+    business_logic::{
+        execution::{
+            execution_entry_point::ExecutionEntryPoint,
+            objects::{CallInfo, TransactionExecutionContext},
+        },
+        fact_state::{
+            in_memory_state_reader::InMemoryStateReader, state::ExecutionResourcesManager,
+        },
+        state::cached_state::CachedState,
+    },
+    definitions::general_config::StarknetGeneralConfig,
+    services::api::contract_class::EntryPointType,
+};
+use felt::{felt_str, Felt};
+use num_traits::ToPrimitive;
+use std::collections::HashMap;
+
+// ----------------------------------------------------------------------------
+/// Transfers the amount actual_fee from the caller account to the sequencer.
+/// Returns the resulting CallInfo of the transfer call.
+
+pub(crate) fn execute_fee_transfer(
+    state: &mut CachedState<InMemoryStateReader>,
+    general_config: StarknetGeneralConfig,
+    tx_context: TransactionExecutionContext,
+    actual_fee: u64,
+) -> Result<CallInfo, TransactionError> {
+    if actual_fee as u64 > tx_context.max_fee {
+        return Err(TransactionError::FeeError(
+            "Actual fee exceeded max fee.".to_string(),
+        ));
+    }
+
+    let fee_token_address = general_config.starknet_os_config.fee_token_address.clone();
+
+    let calldata = [
+        general_config.block_info.sequencer_address.0.clone(),
+        Felt::from(actual_fee),
+        0.into(),
+    ]
+    .to_vec();
+
+    // Value generated from `get_selector_from_name('TRANSFER_ENTRY_POINT_SELECTOR')`.
+    let entry_point_selector =
+        felt_str!("232670485425082704932579856502088130646006032362877466777181098476241604910");
+
+    let fee_transfer_call = ExecutionEntryPoint::new(
+        fee_token_address,
+        calldata,
+        entry_point_selector,
+        tx_context.account_contract_address.clone(),
+        EntryPointType::External,
+        None,
+        None,
+    );
+
+    let mut resources_manager = ExecutionResourcesManager::default();
+    fee_transfer_call
+        .execute(state, &general_config, &mut resources_manager, &tx_context)
+        .map_err(|_| TransactionError::FeeError("Fee transfer failure".to_string()))
+}
+
+// ----------------------------------------------------------------------------------------
+/// Calculates the fee of a transaction given its execution resources.
+/// We add the l1_gas_usage (which may include, for example, the direct cost of L2-to-L1
+/// messages) to the gas consumed by Cairo resource and multiply by the L1 gas price.
+
+pub(crate) fn calculate_tx_fee(
+    resources: HashMap<String, Felt>,
+    gas_price: u64,
+    general_config: &StarknetGeneralConfig,
+) -> Result<u64, TransactionError> {
+    let gas_usage = resources
```

Comment:
This is something that vm outputs? Sounds strange
cc @juanbono @mmsc2

Reply:
the vm outputs the resources and with that we can calculate the gas